### ISSUE TITLE
Conflicts with other plugins that use "/"

### ISSUE THREAD

**Issue by IndianBoy42** [gelguy/wilder.nvim#63]
specifically vim-visual-multi has a feature that allows you to multi select all matches of a certain regex. However with wilder.nvim enabled for '/' the vim-visual-multi regex search gets replaced by the wilder regex search. 

Is there anyway to get these plugins to work together? ideally I could use wilder completions in the vim-visual-multi, but another way could be to have the wilder search only trigger when called from a certain keybinding (ie have a separate function to start wilder search)

**Comment by gelguy** [comment]
To clarify, `wilder` doesn't hijack or remap `/` (unless you use a mapping for `wilder#start_from_normal_mode()`). It uses `CmdlineEnter` to detect that the `/` or `:` cmdline has been entered.

What is happening here is that `vim-visual-multi` remaps `/` and when `multi` mode is entered, the statusline is changed to show additional info for `multi` mode. `wilder` is drawing over the statusline, which is causing the conflict here.

For triggering `wilder` on a keybinding, you can use
```vim
" remove the wilder#enable_cmdline_enter() call
cmap <expr> <F1> wilder#main#start()
```
`wilder` will then start only after `<F1>` is pressed.

This can be followed up with another mapping
```vim
nmap g/ /<F1>
```
which would avoid conflicting with the `/` mapping of `vim-visual-multi`.

**Comment by IndianBoy42** [comment]
Ah, I think I understand now. thank you, I will try your suggestions

**Comment by gelguy** [comment]
`vim-visual-multi` exposes enough information to recreate the statusline so an alternative would be to add that information to `wilder#wildmenu_renderer()`.

Very rough example to demonstrate:
```vim
  let s:Wildmenu_renderer = wilder#wildmenu_renderer(({
        \ 'left': [{-> s:vm()}, wilder#previous_arrow()],
        \ 'right': [wilder#next_arrow()],
        \ }))

  function! s:foo()
    if !exists('b:visual_multi')
      return ''
    endif
    let v = b:VM_Selection.Vars
    let vm = VMInfos()
    let color  = '%#VM_Extend#'

    try
      if v.insert
        if vm.Insert.replace
          let [ mode, color ] = [ 'V-R', '%#VM_Mono#' ]
        else
          let [ mode, color ] = [ 'V-I', '%#VM_Cursor#' ]
        endif
      else
        let mode = { 'n': 'V-M', 'v': 'V', 'V': 'V-L', "\<C-v>": 'V-B' }[mode()]
      endif
    catch
      let mode = 'V-M'
    endtry
    let mode = exists('v.statusline_mode') ? v.statusline_mode : mode
    let patterns = string(vm.patterns)[:(winwidth(0)-30)]

    return printf(' %s %s %s %s ', mode, vm.ratio, patterns, vm.status)
  endfunction
```
For reference: https://github.com/mg979/vim-visual-multi/blob/f994695813ebaecc9e37c7ea216c65d9cd659767/autoload/vm/themes.vim#L100-L126

**Comment by IndianBoy42** [comment]
I'm not sure but I don't think the problem is just the statusline getting covered up, because it also functionally doesn't work.

It reverts back to just searching the document, without doing the vim-visual-multi things.

**Comment by gelguy** [comment]
Can you paste your config here?

**Comment by IndianBoy42** [comment]
My whole config is actually here on github: https://github.com/indianBoy42/lunarvim

the wilder config is in `lua/lv-wilder/init.lua`

**Comment by gelguy** [comment]
Could you also list steps to reproduce?

**Comment by IndianBoy42** [comment]
BTW: disabling `wilder#enable_cmdline_enter()` and enabling wilder separately seems to work, which is what is on the latest commit in the repo

**Comment by IndianBoy42** [comment]
To reproduce the problem 

1. uncomment out `wilder#enable_cmdline_enter()` firstly
2. Visually select some region of test
3. enter `/` to enter vim-visual-multi "Visual Regex"
4. the vim-visual-multi statusline would flicker briefly and the regex search will act like a vanilla vim search
5. if `wilder#enable_cmdline_enter()` is disabled then after step 3, you would be able to type some search and when you hit enter, the found regions will be added as selections in vim-visual-multi

**Comment by gelguy** [comment]
I'm not able to reproduce this - am I missing a keymapping (I'm using the default `g:VM_leader` `\\`)?

1. Visually selecting a region and entering `/` doesn't enter `VM-REGEX` mode, it's Vim's default Visual mode
2. Visually selecting a region and entering `\\/` enters `VM-REGEX` mode
3. `VM-REGEX` mode is the same as searching with `/`, after entering your search with `<CR>`:
a. If there's one or more matches, it goes back to `V-M` mode with the matches added to selections
b. If there's no match across the whole buffer, a `Pattern not found` error is thrown, and it goes back to `VM-REGEX` mode
c. If there's no match within the lines of the visual selection, but a match in the buffer, therer's no error thrown and it goes back to normal mode (likely a bug too?)
d. Note that matches also have to start after the cursor
e. Matches are searched within the lines where the visual selection starts and stops, so it's possible to have a match outside the visual selection if it's on the same line
4. If `wilder` is active, there's no difference in the behavior, other than the statusline being drawn over

Can I confirm my observations are correct?

**Comment by IndianBoy42** [comment]
For point one I have remapped "/" in visual mode to do VM-REGEX, sorry for not mentioning that

point 2 is correct for the default mapping

point 3 is how it should work

point 4 is where we diverge for some reason, after searching it doesn't go to V-M mode, it goes to normal mode for me, which it shouldnt do. This only happens if wilder is active, either from cmdline_enter or from `wilder#main#start()`

also I have noticed that the search highlighting is a bit weird with wilder active, the highlight disappears as I type, although it shows up permanently after I hit `<CR>`. I don't know if this is expected/related

**Comment by gelguy** [comment]
Regarding the highlights, see https://github.com/gelguy/wilder.nvim/issues/30.

Can you check if Step 3c  is being reached? The match has to start after the cursor and be within the lines of the visual selection.

e.g.
```
|f|oobar
```
If `f` is the visual selection, `foo` is not considered a match and it would go back to normal mode. This occurs with or without `wilder`. I'm using a minimal config with `vim-visual-multi` as the only plugin.
```vim
" set to wherever vim-plug was installed
set runtimepath^=~/.vim/

call plug#begin('~/.vim/plugins')
Plug 'mg979/vim-visual-multi'
call plug#end()

let g:VM_maps = {}
let g:VM_maps["Visual Regex"] = "/"
```

**Comment by IndianBoy42** [comment]
No 3c isn't being reached, I am searching for something that definitely has a match in the visual selection region 

I just tried to implement the fix from #30, I changed to use `wildmenu_renderer` instead of `popupmenu_renderer` and now VM-REGEX works. Switching back makes it stop working again

The popupmenu renderer uses a buffer right? to change the contents does vim technically switch out of the file buffer? This would rip us out of any visual-multi mode (V-M or VM-REGEX)

**Comment by IndianBoy42** [comment]
I have encountered this problem of BufLeave/BufExit autocmds being run due to popup windows like that (compe autocompletion windows for example)

**Comment by gelguy** [comment]
I think some other plugin in combination with `wilder` and `vim-visual-multi` is causing this issue.

I don't have a better workaround than the one in https://github.com/gelguy/wilder.nvim/issues/63#issuecomment-894679905.

**Comment by IndianBoy42** [comment]
Did you try wilder+vim-visual-multi using `popupmenu_renderer` or with `wildmenu_renderer`? Can you confirm whether they work together with popupmenu for you?

**Comment by gelguy** [comment]
Both work fine for me, other than the statusline being covered.

**Comment by gelguy** [comment]
I'm closing this issue since the original question is solved by https://github.com/gelguy/wilder.nvim/issues/63#issuecomment-894679905.

If you wish to, we can continue the investigation with `vim-visual-multi` in another issue if you open one.
